src/main/java/org/jenkinsci/plugins/workflow/cps/steps/ingroovy/GroovyStepExecution.java

```diff
@@ -0,0 +1,149 @@
+import org.jenkinsci.plugins.workflow.steps.BodyExecutionCallback;
+import org.jenkinsci.plugins.workflow.steps.StepExecution;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+
+/**
+ * {@link StepExecution} to be implemented in Groovy
+ *
+ * @author Kohsuke Kawaguchi
+ * @see "doc/step-in-groovy.md"
+ */
+public abstract class GroovyStepExecution extends StepExecution {
+    /**
+     * Captures parameters that invoke {@link GroovyStep}
+     */
+    private transient GroovyStep step;
+
+    /**
+     * Represents the currently executing groovy code that defines the step.
+     */
+    private BodyExecution execution;
+
+    /*package*/ CpsStepContext context;
+
+    /*package*/ void init(GroovyStep step, CpsStepContext context) {
+        this.step = step;
+        this.context = context;
+    }
+
+    @Nonnull
+    @Override
+    public CpsStepContext getContext() {
+        return context;
+    }
+
+    /**
+     * Obtains the {@link GroovyStep} that captures parameters given to this step.
+     */
+    public GroovyStep getStep() {
+        return step;
+    }
+
+    @Override
+    public boolean start() throws Exception {
+        CpsStepContext cps = getContext();
+        CpsThread t = CpsThread.current();
+
+        // TODO: make sure this class is actually CPS transformed
+
+        Closure body = InvokerHelper.getMethodPointer(this, "call");
+        if (getStep().getDescriptor().takesImplicitBlockArgument())
+            body = body.curry(new Body());
+
+        execution = cps.newBodyInvoker(t.getGroup().export(body))
+//                .withStartAction(/*... maybe a marker for the future?*/)
+                .withCallback(BodyExecutionCallback.wrap(cps))
+                .start();
+
+        return false;
+    }
+
+    @Override
+    public void stop(@Nonnull Throwable cause) throws Exception {
+        execution.cancel(cause);
+    }
+
+    /**
+     * Persist {@link #step} in a form that's serializable.
+     */
+    private void writeObject(ObjectOutputStream oos) throws IOException {
+        oos.defaultWriteObject();
+        oos.writeObject(UninstantiatedDescribable.from(step));
+    }
+
+    /**
+     * Pair up with {@link #writeObject(ObjectOutputStream)} to restore {@link #step}
+     */
+    private void readObject(ObjectInputStream ois) throws IOException, ClassNotFoundException {
+        ois.defaultReadObject();
+        UninstantiatedDescribable ud = (UninstantiatedDescribable)ois.readObject();
+        try {
+            step = (GroovyStep)ud.instantiate();
+        } catch (Exception e) {
+            throw new IOException("Failed to instantiate "+ud,e);
+        }
+    }
+
+    /**
+     * Attempt to resolve a method call as a step invocation like Pipeline Script.
+     */
+    public Object methodMissing(String name, Object args) throws IOException {
```

Comment:
This method is invoked via convention. I added a link to the Groovy documentation about this.